lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,385 @@
+    if (!selectCaseLogicalValsInsertStatus.second) {
+      ReportOverlapCaseValError(source,
+          selectCaseLogicalValsInsertStatus.first->second,
+          GetValueAsString(logicalVal).c_str());
+    }
+  }
+  template<typename T>
+  void Insert(std::optional<T> lowerVal, std::optional<T> upperVal,
+      parser::CharBlock source) {
+    SelectCaseRangeType<T> rangeValue(lowerVal, upperVal, source);
+    auto &rangeValsHandle{GetRangeValHandle<T>()};
+    auto &valsHandle{GetValHandle<T>()};
+    auto selectCaseValsIter{std::find_if(valsHandle.begin(), valsHandle.end(),
+        [&rangeValue, this](const auto &mem) {
+          return this->CheckOverlappingValue(
+              mem.first, rangeValue.fromVal, rangeValue.toVal);
+        })};
+    if (selectCaseValsIter != valsHandle.end()) {
+      ReportOverlapCaseValError(source, selectCaseValsIter->second,
+          GetValuePairAsString(lowerVal, upperVal).c_str());
+    }
+    auto selectCaseRangeValsIter{std::find_if(rangeValsHandle.begin(),
+        rangeValsHandle.end(), [&rangeValue, this](const auto &mem) {
+          return ((!mem.fromVal && !rangeValue.fromVal) ||
+              (!mem.toVal && !rangeValue.toVal) ||
+              this->CheckOverlappingRangeValue(mem.fromVal, mem.toVal,
+                  rangeValue.fromVal, rangeValue.toVal));
+        })};
+    if (selectCaseRangeValsIter == rangeValsHandle.end()) {
+      rangeValsHandle.push_back(rangeValue);
+    } else {
+      ReportOverlapCaseValError(source, selectCaseRangeValsIter->source,
+          GetValuePairAsString(lowerVal, upperVal).c_str());
+    }
+  }
+  template<typename T> void InsertString(const T &x, parser::CharBlock source) {
+    if (kind_ == 1) {
+      if (auto val{GetStringValue(x)}) {
+        Insert(val.value(), source);
+      }
+    } else if (kind_ == 2) {
+      if (auto val{GetU16StringValue(x)}) {
+        Insert(val.value(), source);
+      }
+    } else if (kind_ == 4) {
+      if (auto val{GetU32StringValue(x)}) {
+        Insert(val.value(), source);
+      }
+    }
+  }
+  template<typename T>
+  void InsertString(const T &lower, const T &upper, parser::CharBlock source) {
+    if (kind_ == 1) {
+      const auto lowerVal = lower ? GetStringValue(lower->thing) : std::nullopt;
+      const auto upperVal = upper ? GetStringValue(upper->thing) : std::nullopt;
+      Insert(lowerVal, upperVal, source);
+    } else if (kind_ == 2) {
+      const auto lowerVal =
+          lower ? GetU16StringValue(lower->thing) : std::nullopt;
+      const auto upperVal =
+          upper ? GetU16StringValue(upper->thing) : std::nullopt;
+      Insert(lowerVal, upperVal, source);
+    } else if (kind_ == 4) {
+      const auto lowerVal =
+          lower ? GetU32StringValue(lower->thing) : std::nullopt;
+      const auto upperVal =
+          upper ? GetU32StringValue(upper->thing) : std::nullopt;
+      Insert(lowerVal, upperVal, source);
+    }
```

Comment:
assert/check here?

Reply:
if GetU32StringValue returns std::nullopt, insertion will fail and we simply return. Should we add assert for this?

Reply:
i mean a default else and an assertion/check for kinds not handled.

Reply:
Thanks for the detailed review Kiran. I will add a check for default else case.
We do not enter this code for types other than Integer, real and bool. If there is a mismatch in kind or type. IsValidSelectCaseType function will issue an error and return to handle next select case.